browser/ui/webui/settings/brave_settings_leo_assistant_handler.cc

```diff
@@ -0,0 +1,141 @@
+
+namespace {
+
+const std::vector<sidebar::SidebarItem>::const_iterator FindAiChatSidebarItem(
+    const std::vector<sidebar::SidebarItem>& items) {
+  return base::ranges::find_if(items, [](const auto& item) {
+    return item.built_in_item_type ==
+           sidebar::SidebarItem::BuiltInItemType::kChatUI;
+  });
+}
+
+bool ShowLeoAssistantIconVisibleIfNot(
+    sidebar::SidebarService* sidebar_service) {
+  const auto hidden_items = sidebar_service->GetHiddenDefaultSidebarItems();
+  const auto item_hidden_iter = FindAiChatSidebarItem(hidden_items);
+
+  if (item_hidden_iter != hidden_items.end()) {
+    sidebar_service->AddItem(*item_hidden_iter);
+    return true;
+  }
+
+  return false;
+}
+
+bool HideLeoAssistantIconIfNot(sidebar::SidebarService* sidebar_service) {
+  const auto visible_items = sidebar_service->items();
+  const auto item_visible_iter = FindAiChatSidebarItem(visible_items);
+
+  if (item_visible_iter != visible_items.end()) {
+    sidebar_service->RemoveItemAt(item_visible_iter - visible_items.begin());
+    return true;
+  }
+
+  return false;
+}
+
+}  // namespace
+
+namespace settings {
+
+BraveLeoAssistantHandler::BraveLeoAssistantHandler() = default;
+BraveLeoAssistantHandler::~BraveLeoAssistantHandler() = default;
+
+void BraveLeoAssistantHandler::RegisterMessages() {
+  profile_ = Profile::FromWebUI(web_ui());
+
+  web_ui()->RegisterMessageCallback(
+      "toggleLeoIcon",
+      base::BindRepeating(&BraveLeoAssistantHandler::HandleToggleLeoIcon,
+                          base::Unretained(this)));                          
+  web_ui()->RegisterMessageCallback(
+      "getLeoIconVisibility",
+      base::BindRepeating(&BraveLeoAssistantHandler::HandleGetLeoIconVisibility,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "resetLeoData",
+      base::BindRepeating(&BraveLeoAssistantHandler::HandleResetLeoData,
+                          base::Unretained(this)));
+}
+
+void BraveLeoAssistantHandler::OnJavascriptAllowed() {
+  sidebar_service_observer_.Reset();
+  sidebar_service_observer_.Observe(
+      sidebar::SidebarServiceFactory::GetForProfile(profile_));
+}
+
+void BraveLeoAssistantHandler::OnJavascriptDisallowed() {
+  sidebar_service_observer_.Reset();
+}
+
+void BraveLeoAssistantHandler::OnItemAdded(const sidebar::SidebarItem& item,
+                                           size_t index) {
+  if (item.built_in_item_type ==
+      sidebar::SidebarItem::BuiltInItemType::kChatUI) {
+    NotifyChatUiChanged(true);
+  }
+}
+
+void BraveLeoAssistantHandler::OnItemRemoved(const sidebar::SidebarItem& item,
+                                             size_t index) {
+  if (item.built_in_item_type ==
+      sidebar::SidebarItem::BuiltInItemType::kChatUI) {
+    NotifyChatUiChanged(false);
+  }
+}
+
+void BraveLeoAssistantHandler::NotifyChatUiChanged(const bool& isLeoVisible) {
+  if (!IsJavascriptAllowed()) {
+    return;
+  }
```

Comment:
Q. Is this assuming that when IsJavascriptAllowed() is false, leo web ui doesn't exist?

Reply:
Yeah, I can see that. What I curious  about is why we return here rather than call AllowJavasceipr() like other places.

Reply:
You are right, I haven't found any causes to not simply add `AllowJavascript();` instead of `IsJavascriptAllowed()` check.

Reply:
We check for `AllowJavascript()` because maybe the frontend JS hasn't been initialized yet, and can't receive messages. We can't call `AllowJavascript()` here because it's c++ -> JS. We only call that when a message is received _from_ JS to c++.

Reply:
Reverted it back

Reply:
> IsJavascriptAllowed() because maybe the frontend initialized yet,

Well, we start to observe sidebar service when OnJavascriptAllowed() is called, I guess that shouldn't happen. Do you think DCHECK(IsJavascriptAllowed()); makes more sense?

Reply:
I guess the DCHECK is acceptable, to make sure noone calls it before. But either way is ok